AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/helper/ValidatorParameterHelper.java

```diff
@@ -211,337 +272,222 @@ public void validate(ExecutableElement executableElement, IsValid valid) {
-		zeroOrOneSpecificParameter(executableElement, ANDROID_SHERLOCK_MENU_ITEM_QUALIFIED_NAMES, valid);
-	}
-
-	public void zeroOrOneIntentParameter(ExecutableElement executableElement, IsValid isValid) {
-		zeroOrOneSpecificParameter(executableElement, CanonicalNameConstants.INTENT, isValid);
+		@Override
+		protected String description() {
+			return "extending " + typeName;
+		}
 	}
 
-	public void zeroOrOneSpecificParameter(ExecutableElement executableElement, String parameterTypeQualifiedName, IsValid valid) {
-		zeroOrOneSpecificParameter(executableElement, Arrays.asList(parameterTypeQualifiedName), valid);
-	}
+	public class AnnotatedWithParameterRequirement extends BaseParameterRequirement {
 
-	public void zeroOrOneSpecificParameter(ExecutableElement executableElement, List<String> parameterTypeQualifiedNames, IsValid valid) {
+		private Class<? extends Annotation> annotationClass;
 
-		zeroOrOneParameter(executableElement, valid);
+		public AnnotatedWithParameterRequirement(Class<? extends Annotation> annotationClass) {
+			this.annotationClass = annotationClass;
+		}
 
-		List<? extends VariableElement> parameters = executableElement.getParameters();
+		@Override
+		public boolean isSatisfied(VariableElement param) {
+			return param.getAnnotation(annotationClass) != null;
+		}
 
-		if (parameters.size() == 1) {
-			VariableElement parameter = parameters.get(0);
-			TypeMirror parameterType = parameter.asType();
-			if (!parameterTypeQualifiedNames.contains(parameterType.toString())) {
-				valid.invalidate();
-				annotationHelper.printAnnotationError(executableElement, "%s can only be used on a method with no parameter or a parameter of type " + parameterTypeQualifiedNames + ", not " + parameterType);
-			}
+		@Override
+		protected String description() {
+			return "annotated with " + annotationClass.getSimpleName();
 		}
 	}
 
-	public void zeroOrOneBundleParameter(ExecutableElement executableElement, IsValid valid) {
-		zeroOrOneSpecificParameter(executableElement, CanonicalNameConstants.BUNDLE, valid);
-	}
+	public class PrimitiveOrWrapperParameterRequirement extends BaseParameterRequirement {
 
-	public void hasOneOrTwoParametersAndFirstIsBoolean(ExecutableElement executableElement, IsValid valid) {
-		List<? extends VariableElement> parameters = executableElement.getParameters();
+		private TypeKind type;
 
-		if (parameters.size() < 1 || parameters.size() > 2) {
-			valid.invalidate();
-			annotationHelper.printAnnotationError(executableElement, "%s can only be used on a method with 1 or 2 parameter, instead of " + parameters.size());
-		} else {
-			VariableElement firstParameter = parameters.get(0);
+		public PrimitiveOrWrapperParameterRequirement(TypeKind type) {
+			this.type = type;
+		}
 
-			TypeKind parameterKind = firstParameter.asType().getKind();
+		@Override
+		protected String description() {
+			return type + " or " + getWrapperType();
```

Comment:
@yDelouis This will print *BOOLEAN or java.lang.Boolean*. I will add a `tolowercase()` for the first, to avoid confusion.